src/ParallelAlgorithms/Amr/Criteria/TruncationError.hpp

```diff
@@ -0,0 +1,212 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <array>
+#include <cstddef>
+#include <pup.h>
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "DataStructures/DataBox/DataBoxTag.hpp"
+#include "Domain/Amr/Flag.hpp"
+#include "Domain/Tags.hpp"
+#include "NumericalAlgorithms/LinearOperators/PowerMonitors.hpp"
+#include "NumericalAlgorithms/Spectral/Mesh.hpp"
+#include "Options/Context.hpp"
+#include "Options/ParseError.hpp"
+#include "Options/String.hpp"
+#include "ParallelAlgorithms/Amr/Criteria/Criterion.hpp"
+#include "Utilities/Algorithm.hpp"
+#include "Utilities/TMPL.hpp"
+
+/// \cond
+template <size_t>
+class ElementId;
+/// \endcond
+
+namespace amr::Criteria {
+
+/*!
+ * \brief Refine the grid towards the target truncation error
+ *
+ * - If any tensor component has a truncation error above the target value, the
+ *   element will be p-refined.
+ * - If all tensor components still satisfy the target even with one mode
+ *   removed, the element will be p-coarsened.
+ *
+ * For details on how the truncation error is computed see
+ * `PowerMonitors::truncation_error`.
+ *
+ * \tparam Dim Spatial dimension of the grid
+ * \tparam TensorTags List of tags of the tensors to be monitored
+ */
+template <size_t Dim, typename TensorTags>
```

Comment:
[optional] Maybe in the future we will want to make it so that we can runtime choose how to deal with the tensor tags? E.g. we might want to monitor the L2 norm over the spacetime metric components, etc. It would also be nice to be able to have some runtime control over which of the tags we monitor, so we don't need to recompile to monitor different tags. E.g. we could just add all evolved and primitive tags to the list and then decide at runtime which we actually deal with. Again, not for this PR, but wanted to mention it and get your thoughts. I could see this being especially useful during testing/debugging times of the criterion.

Reply:
Agreed. Added the runtime control over which variables to monitor. I left out the norms over tensor components for now (not sure if we really need that because monitoring the L2 norm over something like the metric seems pretty much the same as monitoring each component. We can add this later if we want it.)

Reply:
Sounds good. The advantage of monitoring an L2 norm over components would be fewer matrix multiplies :) But we can experiment with this in the futer.